Updated Smmpt to address ARC feedback (#121)
riscv/riscv-smmtt: chapter4.adoc

MPTEs. Each MPTE is a four byte value.

If the MPTE points to the next level of the memory protection table, it is a
non-leaf MPTE else it is a leaf MPTE. A leaf MPTE provides access type
permissions for eight pages in the address range determined by that MPTE.

The format of a non-leaf MPTE entry is shown in <<rv32-mpte>> and that of a leaf
MPTE entry in <<rv32-mpte-leaf>>.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title="`Smmpt34` non-leaf `MPTE`", id="rv32-mpte"]
[wavedrom, ,svg]
....
{reg: [
  {bits:  1,  name: 'V'},
  {bits:  1,  name: 'L=0'},
  {bits:  8,  name: 'RESERVED'},
  {bits:  22, name: 'PPN'},
], config:{lanes: 2, hspace:1024}}
....

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title="`Smmpt34` leaf `MPTE`", id="rv32-mpte-leaf"]
[wavedrom, ,svg]
....
{reg: [
  {bits:  1,  name: 'V'},
  {bits:  1,  name: 'L=1'},
  {bits:  5,  name: 'RESERVED'},
  {bits:  24, name: 'PERMS'},
  {bits:  1,  name: 'N'},
], config:{lanes: 2, hspace:1024}}
....

The `V` bit indicates if the MPTE is valid; if it is 0, all other bits in the
MPTE are don't-cares and may be used freely by software. When the `MPTE` is

> valid, the `RESERVED` field must be 0.

Reviewer: ```suggestion
valid, the bits constituting the `RESERVED` field are reserved for future standard use and, until its use is defined by some standard extension, must be zeroed by software for forward compatibility. If any of these bits are set, an access-fault exception is raised.
```

> valid, the bits constituting the `RESERVED` field are reserved for future standard use and, until its use is defined by some standard extension, must be zeroed by software for forward compatibility. If any of these bits are set, an access-fault exception is raised.

Author: ```suggestion
valid, the bits constituting the `RESERVED` field are reserved for future
standard use and, until its use is defined by some standard extension,
must be zeroed by software for forward compatibility. If any of these
bits are set, an access-fault exception is raised.
```

[[chapter4]]
[[Smmpt]]

> == `Smmpt`: Machine-mode Memory Protection Table (MPT)

Author: ```suggestion
== Machine-mode Memory Protection Table (MPT)
```

Memory Protection table is a radix-tree table used to provide read, write, and
execute permissions for implicit and explicit memory accesses by effective
privilege modes less than M.

> The following MPT extensions are defined:

Author: ack - removed

* Smmpt64 - Page-based 64-bit memory protection system (RV64 only)
* Smmptnapot - NAPOT Access-Type Permission Granularity (RV32 and R64)
* Smmptnlnapot - NAPOT Memory Protection Table (RV64 only)

These MPT extensions depend on `Smsdid`.

=== Smmpt34: Page-Based Memory Protection System

The RV32 Memory Protection Table provides memory access type permissions
for pages in a 34-bit supervisor physical address (SPA) space. The SPA is
partitioned into a page number (`pn`) and a range offset. When the Smmpt34 mode
is selected in the `MODE` field of `mmpt`, the access type permissions for a
supervisor physical address are determined using a two-level memory protection
table.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title="RV32 Supervisor Physical Address", id=rv32-spa]
[wavedrom, ,svg]
....
{reg: [
  {bits:  15, name: 'range offset', attr: '15'},
  {bits:  10, name: 'pn[0]',        attr: '10' },
  {bits:  9,  name: 'pn[1]',        attr: '9'},
], config:{lanes: 1, hspace:1024}}
....

The root memory protection table has 2^9^ memory-protection-table entries
(MPTEs) and the memory protection tables linked to the root table have 2^10^

> MPTEs. Each MPTE is a four byte value.

Reviewer 2: ```suggestion
MPTEs. Each MPTE is a four byte value. The root memory protection table
is 8 KiB instead of the usual 4 KiB and must be aligned to an 8 KiB boundary.
```

[title="`Smmpt34` non-leaf `MPTE`", id="rv32-mpte"]
[wavedrom, ,svg]
....
{reg: [
  {bits:  1,  name: 'V'},
  {bits:  1,  name: 'L=0'},
  {bits:  8,  name: 'RESERVED'},
  {bits:  22, name: 'PPN'},
], config:{lanes: 2, hspace:1024}}
....

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title="`Smmpt34` leaf `MPTE`", id="rv32-mpte-leaf"]
[wavedrom, ,svg]
....
{reg: [
  {bits:  1,  name: 'V'},
  {bits:  1,  name: 'L=1'},
  {bits:  5,  name: 'RESERVED'},
  {bits:  24, name: 'PERMS'},
  {bits:  1,  name: 'N'},
], config:{lanes: 2, hspace:1024}}
....

The `V` bit indicates if the MPTE is valid; if it is 0, all other bits in the
MPTE are don't-cares and may be used freely by software. When the `MPTE` is
valid, the bits constituting the `RESERVED` field are reserved for future
standard use and, until its use is defined by some standard extension,
must be zeroed by software for forward compatibility. If any of these
bits are set, an access-fault exception is raised.

Explicit stores by a hart that update the `V` bit of leaf and/or non-leaf MPTEs
from 0 to 1 and are visible to a hart will eventually become visible within a

> bounded timeframe to subsequent implicit accesses by that hart to such MPTEs.

Author: I can add the following to the NOTE below (modeled after the note in Svvptc), which is what we should require IMO to provide implementation flexibility:

"This behavior relieves M-mode software from executing certain memory-management
instructions, such as MFENCE.SPA or MINVAL.SPA, which would normally be used to
synchronize the hart’s MPT caches when a memory-resident MPTE is changed from
Invalid to Valid. Synchronizing the hart’s MPT caches with other forms of updates
to a memory-resident MPTE, including when a MPTE is changed from Valid to Invalid,
requires the use of suitable memory-management instructions. This behavior
guarantees that a change to a MPTE from Invalid to Valid is made visible within a
bounded time, thereby making the execution of these memory-management
instructions redundant. The performance benefit of eliding these instructions
outweighs the cost of an occasional gratuitous additional access fault that may occur."

> bounded timeframe to subsequent implicit accesses by that hart to such MPTEs.

Reviewer 2: Yes, I think that would be helpful, and sufficient to resolve my concern. That at least makes it explicit that gratuitous access faults may still occur, which we would expect if we are following the behavior of Svvptc.

that provide the Execute(`X`), Write(`W`), and Read(`R`) permissions. The most
significant 3 bits of the range offset identify a page in that range and are
used to identify the access type permission bits for an access to that page.

[[Smmpt-xwr-encoding]]
.Encoding of `MPTE` X/W/R permission tuple
[%autowidth,float="center",align="center",cols="^,^,^,<",options="header"]
|===
| X | W | R | Meaning
| 0 | 0 | 0 | _Reserved for future use._
| 0 | 0 | 1 | Read-only page.
| 0 | 1 | 0 |  _Reserved for future use._
| 0 | 1 | 1 | Read-write page.
| 1 | 0 | 0 | Execute-only page.
| 1 | 0 | 1 |  Read-execute page.
| 1 | 1 | 0 | _Reserved for future use._
| 1 | 1 | 1 | Read-write-execute page.
|===

Access type permissions -- readable, writable, or executble -- are checked
by MPT the same as for VS-stage and G-stage translation. For a memory access
made to support VS-stage or G-stage address translation (such as a read/write to
a VS-level page table or a G-level page table), permissions are checked as
though for an implicit load or store, not for the original access type. However,
any exception is always reported for the original access type (instruction,
load, or store/AMO).

> If Smmptnapot is not implemented, the bit 31 in leaf MPTE remains reserved.

Reviewer 2: editorial:

```suggestion
If Smmptnapot is not implemented, bit 31 in a leaf MPTE remains reserved.
```

[[MPT_ACC_LKUP]]
==== MPT access type permissions lookup process
Access type permissions for a physical address PA are determined as
follows:

1. Let _a_ be `mmpt.ppn` x PAGESIZE, and let _i_ = LEVELS-1 (For Smmpt34,
   PAGESIZE is 2^12^ and LEVELS=2). The `mmpt` register must be active,
   i.e., the effective privilege mode must not be M-mode.

2. Let _mpte_ be the value of the `MPTE` at address _a_ + _pa.pn[i]_
   x MPTESIZE (For Smmpt34, MPTESIZE=4). If accessing _mpte_ violates a PMA
   or PMP check, raise an access-fault exception corresponding to the original
   access type.

3. If _mpte.v_=0, or if any bits or encodings that are reserved for future
   standard use are set within _mpte_, stop and raise an access-fault exception
   corresponding to the original access type.

4. Otherwise, the _mpte_ is valid. If _mpte.L=1_ go to step 5; Otherwise, this
   MPTE is a pointer to the next level of the memory protection table. Let
   _i_=_i_-1. If _i_<0, stop and raise an access-fault exception corresponding
   to the original access type. Otherwise, let _a_ = _mpte.PPN_ x PAGESIZE and
   go to step 2.

5. A leaf _mpte_ has been found. Let _pi_ be the NUMPGINRANGE (For Smmpt34,

>    NUMPGINRANGE is 3) most significant bits of the range offset. Let

Reviewer 2: For i > 0, _pi_ is the NUMPGINRANGE most significant bits of _pa.pn[i-1]_, _not_ the bits from the range offset.

Also, I would recommend renaming this to something like `PGBITSINRANGE`, since it is really the log base 2 of the number of pages in the range.

Author: ack. I will describe it as when the entry at i > 0 is the leaf, the bits pa.pn[i-1] are included in the range offset.

Reviewer 2: Your update looks good to me.

   _XWR=PERMS[pi*3+2:pi*3]_. Determine if the requested memory access is allowed

>    by the _XWR_ bits, given the current privilege mode and MXR field of the

Reviewer 2: ```suggestion
   by the _XWR_ bits, given the effective privilege mode and MXR field of the
```

   `mstatus` register. If not, stop and raise an access-fault exception
   corresponding to the original access type, else the requested memory access
   is allowed.

Implicit accesses to MPT by this process are checked by PMP/Smepmp as implicit
M-mode accesses. The endianness of such implicit accesses is controlled by
`mstatus.MBE`.

=== Smmpt43: Page-Based 43-bit Memory Protection System

> This section describes a page-based memory protection system RV64, which

Reviewer 2: editorial:

```suggestion
This section describes a page-based memory protection system for RV64, which
```

supports a 43-bit supervisor physical address space. The design of Smmpt43
follows the overall scheme of Smmpt34, and this section details only the
differences between the schemes.

[NOTE]
====
Multiple memory protection systems are specified for RV64 to relieve the tension
between providing a large physical address space and minimizing the memory

> protection cost. For many systems, 43-bits of physical address space is ample,

Reviewer 2: For many RV64 systems, 34 bits of physical address space is also ample. We should define a two-level mode for RV64 as well. Unfortunately, the naming could be confusing, as it conflicts with the RV32 Smmpt34.

Author: This can be achieved with one additional 4K page and an entry on that page. I don't think we should add a mode for that.

Reviewer 2: There's also the performance overhead of the extra level of MPT walk. Small/embedded/real-time systems are less likely to have MPTE caches, so the performance hit will be more noticeable.

and so Smmpt43 suffices. Smmpt52 increases the physical address space to 52
bits, but increases the physical memory capacity dedicated to memory protection
tables and the latency of memory protection table traversals. Smmpt64 increases
the physical address space to 64 bits.
====

Smmpt43 supports a 43-bit physical address space, divided into pages. The
Smmpt43 address is partitioned as shown in  <<rv64-spa-Smmpt43>>. The system
physical address must have bits 43 and higher all equal to 0, or else an
access-fault exception corresponding to the original access type will occur. The
Smmpt43 memory protection table has three levels.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title="Smmpt43 Supervisor Physical Address", id=rv64-spa-Smmpt43]
[wavedrom, ,svg]
....
{reg: [
  {bits:  16, name: 'range offset', attr: '16'},
  {bits:  9,  name: 'pn[0]',        attr: '9' },
  {bits:  9,  name: 'pn[1]',        attr: '9'},

>   {bits:  9,  name: 'pn[2]',        attr: '9'},

Reviewer 2: The diagram should include all 64 bits, with the upper bits set to zero per the text above.

```suggestion
  {bits:  9,  name: 'pn[2]',        attr: '9'},
  {bits:  21, name: '0',            attr: '21'},
```

Author: I am reverting this change since this spa is an input into the lookup process and those upper bits may or may not be zero. HW will still fault if they are not, but HW does not force them to be 0 as inputs into the lookup process. similar convention is followed for Sv as well.

exactly the size of a page and must always be aligned to a page boundary.
The physical page number of the root memory protection table is stored in the
`mmpt` register's PPN field.

The format of a non-leaf MPTE entry is shown in <<rv64-mpte>> and that of a leaf
MPTE entry in <<rv64-mpte-leaf>>.

A leaf MPTE provides access type permissions for sixteen pages in the address
range determined by that MPTE.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title="Smmpt43 non-leaf `MPTE` entry", id="rv64-mpte"]
[wavedrom, ,svg]
....
{reg: [
  {bits:   1, name: 'V'},
  {bits:   1, name: 'L=0'},
  {bits:   8, name: 'RESERVED'},
  {bits:  52, name: 'PPN'},
  {bits:   1, name: 'RSV'},
  {bits:   1, name: 'N'},
], config:{lanes: 2, hspace:1024}}
....

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title="Smmpt43 leaf `MPTE` entry", id="rv64-mpte-leaf"]
[wavedrom, ,svg]
....
{reg: [
  {bits:   1, name: 'V'},
  {bits:   1, name: 'L=1'},
  {bits:   8, name: 'RESERVED'},
  {bits:  48, name: 'PERMS'},
  {bits:   5, name: 'RESERVED'},
  {bits:   1, name: 'N'},
], config:{lanes: 2, hspace:1024}}
....

> If Smmptnapot is not implemented, the bit 63 in leaf MPTE remains reserved. If
> Smmptnlnapot is not implemented, the bit 63 in non-leaf MPTE remains reserved.

Reviewer 2: editorial:

```suggestion
If Smmptnapot is not implemented, bit 63 in a leaf MPTE remains reserved. If
Smmptnlnapot is not implemented, bit 63 in a non-leaf MPTE remains reserved.
```

<<MPT_ACC_LKUP>>, except LEVELS equals 3, MPTESIZE equals 8, and NUMPGINRANGE
equals 4.

=== Smmpt52: Page-Based 52-bit Memory Protection System

This section describes a page-based memory protection system for RV64 that
supports 52-bit physical address spaces. It closely follows the design of
Smmpt43, simply adding an additional level of page table, and so this section
only details the differences between the two schemes.

Smmpt52 supports a 52-bit physical address space, divided into pages. The
Smmpt52 address is partitioned as shown in  <<rv64-spa-Smmpt52>>. The system
physical address must have bits 52 and higher all equal to 0, or else an
access-fault exception corresponding to the original access type will occur. The
Smmpt52 memory protection table has four levels.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title="Smmpt52 Supervisor Physical Address", id=rv64-spa-Smmpt52]
[wavedrom, ,svg]
....
{reg: [
  {bits:  16, name: 'range offset', attr: '16'},
  {bits:  9,  name: 'pn[0]',        attr: '9' },
  {bits:  9,  name: 'pn[1]',        attr: '9'},
  {bits:  9,  name: 'pn[2]',        attr: '9'},

>   {bits:  9,  name: 'pn[3]',        attr: '9'},

Reviewer 2: Likewise this diagram should include all 64 bits of the PA:

```suggestion
  {bits:  9,  name: 'pn[3]',        attr: '9'},
  {bits:  12, name: '0',            attr: '12'},
```

], config:{lanes: 1, hspace:1024}}
....

The Smmpt52 memory protection tables contain 2^9^ memory protection table
entries (MPTEs). Each MPTE is an eight byte value. A page table is exactly the
size of a page and must always be aligned to a page boundary. The physical page
number of the root memory protection table is stored in the `mmpt` register's
PPN field.

> The MPTE formats for Smmpt52 is identical to that of Smmpt43.

Reviewer 2: editorial:

```suggestion
The MPTE formats for Smmpt52 are identical to that of Smmpt43.
```

Any level of MPTE may be a leaf MPTE, so in addition to providing access type
permissions to 4 KiB pages, Smmpt52 supports 32 MiB _megapages_, 16 GiB
_gigapages_, 8 TB _terapages_, and 4 PB _petapages_.

The algorithm to determine access type permissions for a page is same as in
<<MPT_ACC_LKUP>>, except LEVELS equals 4, MPTESIZE equals 8, and NUMPGINRANGE
equals 4.

=== Smmpt64: Page-Based 64-bit Memory Protection System

This section describes a page-based memory protection system for RV64 that
supports 64-bit physical address spaces. It closely follows the design of
Smmpt52, simply adding an additional level of page table, and so this section
only details the differences between the two schemes.

Smmpt64 supports a 64-bit physical address space, divided into pages. The
Smmpt64 address is partitioned as shown in  <<rv64-spa-Smmpt64>>. The

> Smmpt64 memory protection table has four levels.

Reviewer 2: ```suggestion
Smmpt64 memory protection table has five levels.
```

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title="Smmpt64 Supervisor Physical Address", id=rv64-spa-Smmpt64]
[wavedrom, ,svg]
....
{reg: [
  {bits:  16, name: 'range offset', attr: '16'},
  {bits:  9,  name: 'pn[0]',        attr: '9' },
  {bits:  9,  name: 'pn[1]',        attr: '9'},
  {bits:  9,  name: 'pn[2]',        attr: '9'},
  {bits:  9,  name: 'pn[3]',        attr: '9'},
  {bits:  12, name: 'pn[4]',        attr: '12'},
], config:{lanes: 1, hspace:1024}}
....

The Smmpt64 root memory protection table contain 2^12^ memory protection
table entries (MPTEs). Each MPTE is an eight bytes value. The root memory
protection table is 32 KiB instead of the usual 4 KiB and must be aligned to a
32 KiB boundary. The physical page number of the root memory protection table is
stored in the `mmpt` register's PPN field.

The non-root Smmpt64 memory protection tables contain 2^9^ MPTEs, eight bytes
each. The non-root memory protection table is exactly the size of a page and
must always be aligned to a page boundary.

> The MPTE formats for Smmpt64 is identical to that of Smmpt52.

Reviewer 2: editorial:

```suggestion
The MPTE formats for Smmpt64 are identical to that of Smmpt52.
```

Any level of MPTE may be a leaf MPTE, so in addition to providing access type
permissions to 4 KiB pages, Smmpt54 supports 32 MiB _megapages_, 16 GiB
_gigapages_, 8 TB _terapages_, 4 PB _petapages_, and 2 EB _exapages_.

The algorithm to determine access type permissions for a page is same as in
<<MPT_ACC_LKUP>>, except LEVELS equals 5, MPTESIZE equals 8, and NUMPGINRANGE
equals 4.

=== Smmptnapot - NAPOT Access-Type Permission Granularity

When `MPTE.L`=1 and `MPTE.N`=1, the leaf `MPTE` represents a memory protection
range that is part of a larger contigous NAPOT memory protection range comprised

> of 512 pages for RV64 and 1024 pages for RV32. All leaf MPTEs at that level of

Reviewer: The merging in TLB is not disallowed when vs and g stage resolve to 64K and the mpt provides a page size of > 4K.

merging becomes disallowed when the effective page size determined by Vs and G stage is smaller than page size in mpt and not vice versa

> of 512 pages for RV64 and 1024 pages for RV32. All leaf MPTEs at that level of

Reviewer 2: It also becomes a problem if the MPTE size is smaller than the VS/G-stage page size, because then the PTW would have to read multiple MPTEs to determine if the entire page can be accessed. But now I realize this means Svnapot's 64K page size will work fine, because one MPTE covers exactly 64K. The hardware will just need logic to compare all of the PERMS entries within the MPTE and split the page if they are inconsistent. So it's only a concern for new NAPOT page sizes larger than 64K, and maybe that's not worth adding complexity to support.

> of 512 pages for RV64 and 1024 pages for RV32. All leaf MPTEs at that level of

Reviewer: Its not a problem when MPTE is smaller - the permissions are checked only for the bytes that are access. Consider a VS stage PTE creating a 1G page and the G stage mapping that 1G region using 256K 4K pages. This does not prevent the VS stage from accessing any of the 1G page. However, usually its not cached as a 1G page in the TLBs since caching as a 1G translation would require caching 256K SPA+G-stage-permission-sets in that TLB entry.

> of 512 pages for RV64 and 1024 pages for RV32. All leaf MPTEs at that level of

Reviewer 2: > the permissions are checked only for the bytes that are access.  

I know that, but if you want to avoid splitting the page at the TLB, you have to know the permissions for the whole page.

In your example, the TLB entry is 4K, because that's the smallest page size in the lookup path. There will be reduced performance due to increased TLB pressure, but that's a software problem: software _should_ have used 1G pages for the G-stage mapping (barring fragmentation), and then the TLB could use a 1G entry.

My claim is that when the NAPOT VM page size is larger than an MPTE, the TLB entry for that page will _necessarily_ be split, because the "smallest page size in the lookup path" will always be smaller than the NAPOT VM page size, thus negating the performance benefits of using the NAPOT VM page -- that's the problem I am referring to. If I allocate 256K to a supervisor domain, there's no way to make that one NAPOT page in the MPT, so no matter what software does, it cannot get full performance from a hypothetical 256K NAPOT VM page.

> of 512 pages for RV64 and 1024 pages for RV32. All leaf MPTEs at that level of

Reviewer: When a 256K NAPOT page gets defined then along with such a future extension there may be an extension to Smmpt.

> of 512 pages for RV64 and 1024 pages for RV32. All leaf MPTEs at that level of

Reviewer 2: OK, that makes sense. We would use some of the reserved bits to define a new NAPOT MPT page size if a new VM page size is defined.

this NAPOT memory protection range have the same value for L, N, PERMS and V
bits. Further, the XWR access type permssion tuples in the PERMS field of such

> PTEs are identical.

Author: I replaced "Further" with "Thus" - good to call this out explicitly.

> bits. Further, the XWR access type permssion tuples in the PERMS field of such
> PTEs are identical.

Author: ```suggestion
bits. Thus, the XWR access type permission tuples in the PERMS field of such
PTEs are identical.
```

Such MPTEs behave identically to leaf MPTEs in the MPT access permission lookup

> process described in <<MPT-lookup>>, except that:

Reviewer 2: broken link:

```suggestion
process described in <<MPT_ACC_LKUP>>, except that:
```

* For RV32, the `N` bit may be set to 1 in the level 0 leaf MPTEs.

> * For RV64, the `N` bit may be set to 1 in the level 0 and level 1 leaf MPTEs.

Reviewer 2: Why are leaf NAPOT MPTEs not allowed at all levels? Or all non-root levels? Sv48 and Sv57 have larger leaf page sizes.

Author: I think we cover all the cases expected currently? other Napot encodings can be added in the future.

[NOTE]
====
The motivation of NAPOT MPTE is that access-type permissions for one or more
entries representing the contigous region may be cached as a single cache entry
as if it were a single (large) page. This compaction can relieve access-type
permission caches. The scheme allows an implementation to not take advantage of

> this property and simply cache the access-type permissions for pages seperately.

Reviewer 2: This is a good point. Implementations can completely ignore the N bit if they want to. So there's no hardware cost for implementing Smmptnapot. I don't think it should be a separate extension. Svnapot had to be separate because the N bit in PTEs was previously reserved. But since we are defining the N bit in MPTEs from the start, there are no existing Smmpt implementations to be confused by it.

Author: i think it is still good to keep this modular - this aspect is being discussed in the other PR you submitted so will resolve this comment.

Author: non-0 benefit to validation and verification.

Reviewer 2: The only required change on the hardware side is not forcing an access fault in the case L=1 and N=1.

On the other hand, making the extension optional adds cost (to development, validation, and verification) to any software that wants to use leaf NAPOT pages. I would argue that the software cost outweighs the hardware validation cost. Optional extensions aren't free.

Author: I am ok making SmmptNapot a requirement for the Smmpt modes.

[wavedrom, ,svg]
....
{reg: [
  {bits:  1,  name: 'V'},
  {bits:  1,  name: 'L=0'},
  {bits:  8,  name: 'RESERVED'},
  {bits:  22, name: 'PPN'},
], config:{lanes: 2, hspace:1024}}
....

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title="`Smmpt34` leaf `MPTE`", id="rv32-mpte-leaf"]
[wavedrom, ,svg]
....
{reg: [
  {bits:  1,  name: 'V'},
  {bits:  1,  name: 'L=1'},
  {bits:  5,  name: 'RESERVED'},
  {bits:  24, name: 'PERMS'},
  {bits:  1,  name: 'N'},
], config:{lanes: 2, hspace:1024}}
....

The `V` bit indicates if the MPTE is valid; if it is 0, all other bits in the
MPTE are don't-cares and may be used freely by software. When the `MPTE` is
valid, the bits constituting the `RESERVED` field are reserved for future
standard use and, until its use is defined by some standard extension,
must be zeroed by software for forward compatibility. If any of these
bits are set, an access-fault exception is raised.

Explicit stores by a hart that update the `V` bit of leaf and/or non-leaf MPTEs
from 0 to 1 and are visible to a hart will eventually become visible within a
bounded timeframe to subsequent implicit accesses by that hart to such MPTEs.

> [NOTE]
> ====
> Depending on the microarchitecture, some possible ways to facilitate include:
> not storing Invalid PTEs in the MPT page-walk caches, automatically evicting
> Invalid MPTEs using a bounded timer, or making MPT page-walk caches coherent
> with store instructions that modify MPTEs.
> ====

Author: ```suggestion
[NOTE]
====
This behavior relieves M-mode software from executing certain memory-management
instructions, such as MFENCE.SPA or MINVAL.SPA, which would normally be used to
synchronize the hart’s MPT caches when a memory-resident MPTE is changed from
Invalid to Valid. Synchronizing the hart’s MPT caches with other forms of updates
to a memory-resident MPTE, including when a MPTE is changed from Valid to Invalid,
requires the use of suitable memory-management instructions. This behavior
guarantees that a change to a MPTE from Invalid to Valid is made visible within a
bounded time, thereby making the execution of these memory-management
instructions redundant. The performance benefit of eliding these instructions
outweighs the cost of an occasional gratuitous additional access fault that may occur.

Depending on the microarchitecture, some possible ways to facilitate include:
not storing Invalid PTEs in the MPT page-walk caches, automatically evicting
Invalid MPTEs using a bounded timer, or making MPT page-walk caches coherent
with store instructions that modify MPTEs.
====
```

follows the overall scheme of Smmpt34, and this section details only the
differences between the schemes.

[NOTE]
====
Multiple memory protection systems are specified for RV64 to relieve the tension
between providing a large physical address space and minimizing the memory
protection cost. For many systems, 43-bits of physical address space is ample,
and so Smmpt43 suffices. Smmpt52 increases the physical address space to 52
bits, but increases the physical memory capacity dedicated to memory protection
tables and the latency of memory protection table traversals. Smmpt64 increases
the physical address space to 64 bits.
====

Smmpt43 supports a 43-bit physical address space, divided into pages. The
Smmpt43 address is partitioned as shown in  <<rv64-spa-Smmpt43>>. The system
physical address must have bits 43 and higher all equal to 0, or else an
access-fault exception corresponding to the original access type will occur. The
Smmpt43 memory protection table has three levels.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title="Smmpt43 Supervisor Physical Address", id=rv64-spa-Smmpt43]
[wavedrom, ,svg]
....
{reg: [

>   {bits:  16, name: 'range offset', attr: '16'},
>   {bits:  9,  name: 'pn[0]',        attr: '9' },
>   {bits:  9,  name: 'pn[1]',        attr: '9'},
>   {bits:  9,  name: 'pn[2]',        attr: '9'},
>   {bits:  21, name: '0',            attr: '21'},

Author: ```suggestion
  {bits:  16, name: 'range offset', attr: '16'},
  {bits:  9,  name: 'pn[0]',        attr: '9' },
  {bits:  9,  name: 'pn[1]',        attr: '9'},
  {bits:  9,  name: 'pn[2]',        attr: '9'},
```

must always be aligned to a page boundary.

The MPTE formats for Smmpt64 are identical to that of Smmpt52.

Any level of MPTE may be a leaf MPTE, so in addition to providing access type
permissions to 4 KiB pages, Smmpt54 supports 32 MiB _megapages_, 16 GiB
_gigapages_, 8 TB _terapages_, 4 PB _petapages_, and 2 EB _exapages_.

The algorithm to determine access type permissions for a page is same as in
<<MPT_ACC_LKUP>>, except LEVELS equals 5, MPTESIZE equals 8, and NUMPGINRANGE
equals 4.

=== Smmptnapot - NAPOT Access-Type Permission Granularity

When `MPTE.L`=1 and `MPTE.N`=1, the leaf `MPTE` represents a memory protection
range that is part of a larger contiguous NAPOT memory protection range comprised
of 512 pages for RV64 and 1024 pages for RV32. All leaf MPTEs at that level of
this NAPOT memory protection range have the same value for L, N, PERMS and V
bits. Thus, the XWR access type permission tuples in the PERMS field of such
PTEs are identical.

Such MPTEs behave identically to leaf MPTEs in the MPT access permission lookup
process described in <<MPT_ACC_LKUP>>, except that:

* For RV32, the `N` bit may be set to 1 in the level 0 leaf MPTEs.

> * For RV64, the `N` bit may be set to 1 in the level 0 and level 1 leaf MPTEs.

Author: ```suggestion
* For RV64, the `N` bit may be set to 1 in the leaf MPTEs.
```

The scheme supports caching a single access-type permission entry representing a
2 MiB or a 1 GiB memory range for RV64. For RV32, it supports caching a single
access-type permission entry representing a 4 MiB memory range. These contigous
memory range sizes represent large/huge page sizes used commonly by memory
allocators.

A single MPTE entry covers a 64 KiB address range, so a separate 64 KiB NAPOT
PTE is not needed. One MPTE with all XWR permissions being identical provides
the 64 KiB region support.
====

If Smmptnapot extension is implemented then one of Smmpt34, Smmpt43, Smmpt52,
or Smmpt64 extensions must be implemented.

=== Smmptnlnapot - NAPOT Memory Protection Table

The Smmptnlnapot extension supports NAPOT non-leaf MPTEs. This extension is
RV64 only, and depends on one of Smmpt43, Smmpt52 or Smmpt64.

For RV64, when the `MPTE.L`=0 and `MPTE.N`=1, the non-leaf `MPTE`
is part of a range of MPTEs at that level of MPT with the same value for
the L, N, PPN, and V bits.

The following Non-leaf `MPTE` encodings are defined when _mpte_.N is 1.

[[Smmpt-napot]]
.Non-leaf `MPTE` encodings when _mpte_.N=1
[width="100%",cols="10%,30%,30%,30%",options="header",]
|===

> | *_i_* a| *_mpte.PPN[8:0]_* a| *Description* a| *_mpte.napot_bits_*
> |    0  a|       x xxxx xxxx a| Reserved      a| -
> |  > 0  a|       x 1xxx xxxx a| Reserved      a| -
> |  > 0  a|       1 0000 0000 a| 2 MiB contiguous next-level MPT a| 9
> |  > 0  a|       0 xxxx xxxx a| Reserved      a| -

Author: ```suggestion
| *_i_* a| *_mpte.PPN[8:0]_* a| *Description* a| *_mpte.napot_bits_*
|    0  a|       x xxxx xxxx a| Reserved      a| -
|  > 0  a|       x 1xxx xxxx a| Reserved      a| -
|  > 0  a|       x x1xx xxxx a| Reserved      a| -
|  > 0  a|       x Xx1x xxxx a| Reserved      a| -
|  > 0  a|       x xxx1 xxxx a| Reserved      a| -
|  > 0  a|       x xxxx 1xxx a| Reserved      a| -
|  > 0  a|       x xxxx x1xx a| Reserved      a| -
|  > 0  a|       x xxxx xx1x a| Reserved      a| -
|  > 0  a|       x xxxx xxx1 a| Reserved      a| -
|  > 0  a|       1 0000 0000 a| 2 MiB contiguous next-level MPT a| 9
|  > 0  a|       0 xxxx xxxx a| Reserved      a| -
```

The scheme supports caching a single access-type permission entry representing a
2 MiB or a 1 GiB memory range for RV64. For RV32, it supports caching a single
access-type permission entry representing a 4 MiB memory range. These contigous
memory range sizes represent large/huge page sizes used commonly by memory
allocators.

A single MPTE entry covers a 64 KiB address range, so a separate 64 KiB NAPOT
PTE is not needed. One MPTE with all XWR permissions being identical provides
the 64 KiB region support.
====

If Smmptnapot extension is implemented then one of Smmpt34, Smmpt43, Smmpt52,
or Smmpt64 extensions must be implemented.

=== Smmptnlnapot - NAPOT Memory Protection Table

The Smmptnlnapot extension supports NAPOT non-leaf MPTEs. This extension is
RV64 only, and depends on one of Smmpt43, Smmpt52 or Smmpt64.

For RV64, when the `MPTE.L`=0 and `MPTE.N`=1, the non-leaf `MPTE`
is part of a range of MPTEs at that level of MPT with the same value for
the L, N, PPN, and V bits.

The following Non-leaf `MPTE` encodings are defined when _mpte_.N is 1.

[[Smmpt-napot]]
.Non-leaf `MPTE` encodings when _mpte_.N=1
[width="100%",cols="10%,30%,30%,30%",options="header",]
|===

> | *_i_* a| *_mpte.PPN[i]_*  a| *Description*        a|*_mpte.napot_bits_*
> | < 2   a| x xxxx xxxx xxxx a| Reserved             a| -
> |   2   a| x xxx1 0000 0000 a| 2 MiB contiguous L2  a| 9
> |   2   a| x xxx0 xxxx xxxx a| Reserved             a| -
> |   3   a| x xxx1 0000 0000 a| 2 MiB L3             a| 9
> |   3   a| x xxx0 xxxx xxxx a| Reserved             a| -
> |   4   a| x 1000 0000 0000 a| 16 MiB contiguous L4 a| 12
> |   4   a| x 0xxx xxxx xxxx a| Reserved             a| -

Reviewer 2: ```suggestion
| *_i_* a| *_mpte.PPN[0]_* a| *Description*        a| *_mpte.napot_bits_*
|    0  a|     x xxxx xxxx a| Reserved             a| -
|  > 0  a|     x xxxx xxx1 a| Reserved             a| -
|  > 0  a|     x xxxx xx1x a| Reserved             a| -
|  > 0  a|     x xxxx x1xx a| Reserved             a| -
|  > 0  a|     x xxxx 1xxx a| Reserved             a| -
|  > 0  a|     x xxx1 xxxx a| Reserved             a| -
|  > 0  a|     x xx1x xxxx a| Reserved             a| -
|  > 0  a|     x x1xx xxxx a| Reserved             a| -
|  > 0  a|     x 1xxx xxxx a| Reserved             a| -
|  > 0  a|     1 0000 0000 a| 2 MiB next-level MPT a| 9
|  > 0  a|     0 xxxx xxxx a| Reserved             a| -
```

* The first row indicates that non-leaf NAPOT MPTEs are disallowed in the level 0 (leaf) MPT.
* The middle rows indicate that NAPOT sizes smaller than 2 MiB are reserved.
* The second to last row is for the allowed encoding that points to a 2 MiB *next* level table. The size of the table containing the non-leaf NAPOT MPTE only matters in that there must be space for 512 MPTEs. This is always the case for the RV64 Smmpt modes, so no other conditions need to be stated.
* The last row indicates that NAPOT sizes greater than 2 MiB are reserved.

If you really want a Smmtt64 L4 (root) MPTE to contain a single set of NAPOT MPTEs, instead of 8, you would need another table/row to indicate that for _i_ = 4, the combination `mpte.PPN[1] = x xxxx x100` and `mpte.PPN[0] = 0 0000 0000` is allowed. But as I mentioned on #126, I don't think that's necessary.

Author: see below for shortened table

> [[Smmpt-napot]]
> .Non-leaf `MPTE` encodings when _mpte_.N=1
> [width="100%",cols="10%,30%,30%,30%",options="header",]
> |===
> | *_i_* a| *_mpte.PPN[i]_*  a| *Description*        a|*_mpte.napot_bits_*
> | < 2   a| x xxxx xxxx xxxx a| Reserved             a| -
> |   2   a| x xxx1 0000 0000 a| 2 MiB contiguous L2  a| 9
> |   2   a| x xxx0 xxxx xxxx a| Reserved             a| -
> |   3   a| x xxx1 0000 0000 a| 2 MiB L3             a| 9
> |   3   a| x xxx0 xxxx xxxx a| Reserved             a| -
> |   4   a| x 1000 0000 0000 a| 16 MiB contiguous L4 a| 12
> |   4   a| x 0xxx xxxx xxxx a| Reserved             a| -

Author: ```suggestion
[[Smmpt-napot]]
.Non-leaf `MPTE` encodings when _mpte_.N=1
[width="100%",cols="10%,30%,30%,30%",options="header",]
|===
| *_i_* a| *_mpte.PPN[8:0]_* a| *Description* a| *_mpte.napot_bits_*
|    0  a|       x xxxx xxxx a| Reserved      a| -
|  > 0  a|       x 1xxx xxxx a| Reserved      a| -
|  > 0  a|       1 0000 0000 a| 2 MiB contiguous next-level MPT a| 9
|  > 0  a|       0 xxxx xxxx a| Reserved      a| -
```

Author: we could just state it this way then

|===

Such MPTEs behave identically to non-leaf MPTEs in the MPT access permission
lookup process described in <<MPT-lookup>>, except that:

 * If the encoding in _mpte_ is valid according to table <<Smmpt-napot>>,
   then instead of returning the original value of the _mpte_, implicit read
   of a non-leaf NAPOT _mpte_ returns a copy of `mpte` in which

>    _mpte.ppn[i][mpte.napot_bits-1:0]_ are replaced by
>    _pn[i][mpte.napot_bits-1:0]_. If the encoding in _mpte_ is reserved

Reviewer 2: ```suggestion
   _mpte.ppn[mpte.napot_bits-1:0]_ are replaced by
   _pa.pn[i][mpte.napot_bits-1:0]_. If the encoding in _mpte_ is reserved
```

>> It is always mpte.ppn[0] that is overwritten, not mpte.ppn[i]
> 
> That does not seem right

OK, I think the confusion is because _mpte.ppn_ is not split into fields for each MPT level -- it's just one 52 bit field. This is different from Svnapot, where _pte.ppn_ is split into 9-bit subfields.

For clarity, apologies if this is obvious: A non-leaf NAPOT MPTE points to a 2 MiB page (21 bit offset) made up of 512 x 4 KiB pages (9 bit page number + 12 bit page offset) It's these 9 bits that vary based on which of the NAPOT MPTEs is being traversed -- bits 20:12 of the _next_ MPTE's physical address, or in other words _mpte.ppn[8:0]_. Regardless of which level MPT contains the NAPOT MPTE, it's always the same bits that vary. Consider that if _mpte.ppn[17:9]_ were overwritten, the next-level MPT would not be contiguous! It would be 512 4 KiB pages, each 2 MiB apart from each other, covering a total of 1 GiB of address space.

The way Svnapot is written _looks_ like it would extend to i > 0, but does not. Non-leaf Svnapot would similarly need to always replace _pte.ppn[0]_, not the higher bits.

Reviewer 2: It looks like this fix didn't get applied -- only the second bullet got changed.

   according to <<Smmpt-napot>> then an access-fault exception corresponding to
   the original access type must be raised.
 * Implicit reads of non-leaf NAPOT MPTEs may create MPT walk cache entries
   mapping _a_ + _j_*PTESIZE to a copy of _mpte_ in which

>    _mpte[i][pte.napot_bits-1:0]_ is replaced by _pn[i][mpte.napot_bits-1:0]_,
>    for any and all _j_ such that _j_ >> _napot_bits_ = _pn[i].napot_bits_.

Reviewer 2: ```suggestion
   _mpte[mpte.napot_bits-1:0]_ is replaced by _pa.pn[i][mpte.napot_bits-1:0]_,
   for any and all _j_ such that _j_ >> _napot_bits_ = _pa.pn[i]_ >> _napot_bits_.
```

* Same fix as above
* Copy/paste issues from Svnapot

Author: fixed a typoe _mpte -> _mpte.ppn below

>    _mpte[i][pte.napot_bits-1:0]_ is replaced by _pn[i][mpte.napot_bits-1:0]_,
>    for any and all _j_ such that _j_ >> _napot_bits_ = _pn[i].napot_bits_.

Author: ```suggestion
   _mpte.ppn[mpte.napot_bits-1:0]_ is replaced by _pa.pn[i][mpte.napot_bits-1:0]_,
   for any and all _j_ such that _j_ >> _napot_bits_ = _pa.pn[i]_ >> _napot_bits_.
```

[NOTE]
====

> Some implementations may cache non-leaf MPTE to accelerate the MPT access-type

Reviewer 2: editorial:

```suggestion
Some implementations may cache non-leaf MPTEs to accelerate the MPT access-type
```

The scheme supports caching a single access-type permission entry representing a
2 MiB or a 1 GiB memory range for RV64. For RV32, it supports caching a single
access-type permission entry representing a 4 MiB memory range. These contigous
memory range sizes represent large/huge page sizes used commonly by memory
allocators.

A single MPTE entry covers a 64 KiB address range, so a separate 64 KiB NAPOT
PTE is not needed. One MPTE with all XWR permissions being identical provides
the 64 KiB region support.
====

If Smmptnapot extension is implemented then one of Smmpt34, Smmpt43, Smmpt52,
or Smmpt64 extensions must be implemented.

=== Smmptnlnapot - NAPOT Memory Protection Table

The Smmptnlnapot extension supports NAPOT non-leaf MPTEs. This extension is
RV64 only, and depends on one of Smmpt43, Smmpt52 or Smmpt64.

For RV64, when the `MPTE.L`=0 and `MPTE.N`=1, the non-leaf `MPTE`
is part of a range of MPTEs at that level of MPT with the same value for
the L, N, PPN, and V bits.

The following Non-leaf `MPTE` encodings are defined when _mpte_.N is 1.

[[Smmpt-napot]]
.Non-leaf `MPTE` encodings when _mpte_.N=1
[width="100%",cols="10%,30%,30%,30%",options="header",]
|===

> | *_i_* a| *_mpte.PPN[8:0]_* a| *Description* a| *_mpte.napot_bits_*
> |    0  a|       x xxxx xxxx a| Reserved      a| -
> |  > 0  a|       x 1xxx xxxx a| Reserved      a| -
> |  > 0  a|       x x1xx xxxx a| Reserved      a| -
> |  > 0  a|       x Xx1x xxxx a| Reserved      a| -
> |  > 0  a|       x xxx1 xxxx a| Reserved      a| -
> |  > 0  a|       x xxxx 1xxx a| Reserved      a| -
> |  > 0  a|       x xxxx x1xx a| Reserved      a| -
> |  > 0  a|       x xxxx xx1x a| Reserved      a| -
> |  > 0  a|       x xxxx xxx1 a| Reserved      a| -
> |  > 0  a|       1 0000 0000 a| 2 MiB contiguous next-level MPT a| 9
> |  > 0  a|       0 xxxx xxxx a| Reserved      a| -

Author: ```suggestion
| *_i_* a| *_mpte.PPN[8:0]_* a| *Description* a| *_mpte.napot_bits_*
|    0  a|       x xxxx xxxx a| Reserved      a| -
|  > 0  a|       x xxxx xxx1 a| Reserved      a| -
|  > 0  a|       x xxxx xx1x a| Reserved      a| -
|  > 0  a|       x xxxx x1xx a| Reserved      a| -
|  > 0  a|       x xxxx 1xxx a| Reserved      a| -
|  > 0  a|       x xxx1 xxxx a| Reserved      a| -
|  > 0  a|       x xx1x xxxx a| Reserved      a| -
|  > 0  a|       x x1xx xxxx a| Reserved      a| -
|  > 0  a|       x 1xxx xxxx a| Reserved      a| -
|  > 0  a|       1 0000 0000 a| 2 MiB contiguous next-level MPT a| 9
|  > 0  a|       0 xxxx xxxx a| Reserved      a| -
```

[[Smmpt]]
== `Smmpt`: Hart MPT Extension

[[Smmpt-rwx]]
=== Smmpt

> The Supervisor Physical Address (SPA) for RV32 and RV64 is shown below.

Reviewer: ```suggestion
The Supervisor Physical Address (SPA) for RV32 and RV64 are partitioned into a Page Number (pn)  and a page offset as shown in <<add figure reference>> and <<add figure reference>> respectively.
```
